server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,321 @@
+
+    @Inject
+    public DecommissionService(
+        Settings settings,
+        ClusterSettings clusterSettings,
+        ClusterService clusterService,
+        TransportService transportService,
+        ThreadPool threadPool,
+        AllocationService allocationService
+    ) {
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.threadPool = threadPool;
+        this.nodeRemovalExecutor = new NodeRemovalClusterStateTaskExecutor(allocationService, logger);
+        this.decommissionHelper = new DecommissionHelper(
+            clusterService,
+            nodeRemovalExecutor
+        );
+        this.awarenessAttributes = AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(
+            AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING,
+            this::setAwarenessAttributes
+        );
+    }
+
+    List<String> getAwarenessAttributes() {
+        return awarenessAttributes;
+    }
+
+    private void setAwarenessAttributes(List<String> awarenessAttributes) {
+        this.awarenessAttributes = awarenessAttributes;
+    }
+
+    public void initiateAttributeDecommissioning(
+        final DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener,
+        ClusterState state
+    ) {
+        /*
+         * 1. Abdicate master
+         * 2. Register attribute -> status should be set to INIT
+         * 3. Trigger weigh away for graceful decommission -> status should be set to DECOMMISSIONING
+         * 4. Once zone is weighed away -> trigger zone decommission using executor -> status should be set to DECOMMISSIONED on successful response
+         * 5. Clear voting config
+         */
+        this.clusterState = state;
+        logger.info("initiating awareness attribute [{}] decommissioning", decommissionAttribute.toString());
+        excludeDecommissionedClusterManagerNodesFromVotingConfig(decommissionAttribute, listener);
+    }
+
+    private void excludeDecommissionedClusterManagerNodesFromVotingConfig(
+        DecommissionAttribute decommissionAttribute,
+        final ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        final Predicate<DiscoveryNode> shouldAbdicatePredicate = discoveryNode -> nodeHasDecommissionedAttribute(discoveryNode, decommissionAttribute);
+        List<String> clusterManagerNodesToBeDecommissioned = new ArrayList<>();
+        Iterator<DiscoveryNode> clusterManagerNodesIter = clusterState.nodes().getClusterManagerNodes().valuesIt();
+        while (clusterManagerNodesIter.hasNext()) {
+            final DiscoveryNode node = clusterManagerNodesIter.next();
+            if (shouldAbdicatePredicate.test(node)) {
+                clusterManagerNodesToBeDecommissioned.add(node.getName());
+            }
+        }
+        transportService.sendRequest(
```

Comment:
Where would we add the validations to ensure there is no quorum loss due to this ?

Reply:
Added checks for this before we initiate decommissioning - 

```
private static void ensureNoQuorumLossDueToDecommissioning(
        DecommissionAttribute decommissionAttribute,
        Set<DiscoveryNode> clusterManagerNodesToBeDecommissioned,
        CoordinationMetadata.VotingConfiguration votingConfiguration
    ) {
        final Set<String> nodesInVotingConfig = votingConfiguration.getNodeIds();
        assert nodesInVotingConfig.isEmpty() == false;
        final int requiredNodes = nodesInVotingConfig.size() / 2 + 1;
        final Set<String> realNodeIds = new HashSet<>(nodesInVotingConfig);
        realNodeIds.removeIf(ClusterBootstrapService::isBootstrapPlaceholder);
        if (realNodeIds.size() - clusterManagerNodesToBeDecommissioned.size() < requiredNodes) {
            throw new DecommissioningFailedException(
                decommissionAttribute,
                "cannot proceed with decommission request. Cluster might go into quorum loss"
            );
        }
    }
```